File: backend/mcp-code-intelligence-python/src/mcp_code_intel/context/budget_allocator.py
"""Budget Allocator — allocates token budget across ranked results. KSA-171.

Top results get full source, middle get signatures, bottom get references.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Any

from .token_budget_manager import TokenBudgetManager
from .types import MergedResult


@dataclass
class AllocatedResult:
    """A merged result with budget allocation metadata."""

    name: str
    id: int | None = None
    kind: str | None = None
    file: str | None = None
    line: int | None = None
    signature: str | None = None
    source_code: str | None = None
    content: str = ""
    relevance_score: float = 0.0
    sources: list[str] = field(default_factory=list)
    relationship: str | None = None
    detail: str = "reference"  # 'full' | 'signature' | 'reference'
    tokens: int = 0


class BudgetAllocator:
    """Allocates token budget across ranked results with progressive detail."""

    def __init__(self) -> None:
        self._budget_mgr = TokenBudgetManager(4000)

    def allocate(self, results: list[MergedResult], max_tokens: int) -> list[AllocatedResult]:
        """Allocate token budget across merged results."""
        allocated: list[AllocatedResult] = []
        tokens_used = 100  # Response overhead

        high_threshold = max(1, math.ceil(len(results) * 0.2))
        med_threshold = math.ceil(len(results) * 0.6)

        for i, result in enumerate(results):
            if tokens_used >= max_tokens:
                break

            if i < high_threshold:
                detail = "full"
                content = result.source_code or result.content or result.signature or result.name
                tokens = self._budget_mgr.count_tokens(content)
            elif i < med_threshold:
                detail = "signature"
                content = result.signature or result.name
                tokens = self._budget_mgr.count_tokens(content)
            else:
                detail = "reference"
                content = f"{result.name} ({result.file or 'unknown'}:{result.line or 0})"
                tokens = 15

            # Downgrade if exceeds budget
            if tokens_used + tokens > max_tokens and detail == "full":
                detail = "signature"
                content = result.signature or result.name
                tokens = self._budget_mgr.count_tokens(content)

            if tokens_used + tokens <= max_tokens:
                allocated.append(AllocatedResult(
                    name=result.name,
                    id=result.id,
                    kind=result.kind,
                    file=result.file,
                    line=result.line,
                    signature=result.signature,
                    source_code=result.source_code,
                    content=content or "",
                    relevance_score=result.relevance_score,
                    sources=result.sources,
                    relationship=result.relationship,
                    detail=detail,
                    tokens=tokens,
                ))
                tokens_used += tokens

        return allocated

File: backend/mcp-code-intelligence-nodejs/src/context/budget-allocator.ts
/**
 * KSA-160: Budget Allocator — allocates token budget across ranked results.
 * Top results get full source, middle get signatures, bottom get references.
 */

import { MergedResult, ContextItem } from './types.js';

export interface AllocatedResult extends MergedResult {
  detail: 'full' | 'signature' | 'reference';
  content: string;
  tokens: number;
}

export class BudgetAllocator {
  private CHARS_PER_TOKEN = 4;

  /** Allocate token budget across merged results with progressive detail levels. */
  allocate(results: MergedResult[], maxTokens: number): AllocatedResult[] {
    const allocated: AllocatedResult[] = [];
    let tokensUsed = 100; // Response overhead

    const highThreshold = Math.max(1, Math.ceil(results.length * 0.2));
    const medThreshold = Math.ceil(results.length * 0.6);

    for (let i = 0; i < results.length; i++) {
      if (tokensUsed >= maxTokens) break;

      const result = results[i];
      let detail: 'full' | 'signature' | 'reference';
      let content: string;
      let tokens: number;

      if (i < highThreshold) {
        detail = 'full';
        content = result.source_code || result.content || result.signature || result.name;
        tokens = this.estimateTokens(content);
      } else if (i < medThreshold) {
        detail = 'signature';
        content = result.signature || result.name;
        tokens = this.estimateTokens(content);
      } else {
        detail = 'reference';
        content = `${result.name} (${result.file || 'unknown'}:${result.line || 0})`;
        tokens = 15;
      }

      // Downgrade if exceeds budget
      if (tokensUsed + tokens > maxTokens && detail === 'full') {
        detail = 'signature';
        content = result.signature || result.name;
        tokens = this.estimateTokens(content);
      }

      if (tokensUsed + tokens <= maxTokens) {
        allocated.push({ ...result, detail, content, tokens });
        tokensUsed += tokens;
      }
    }

    return allocated;
  }

  private estimateTokens(text: string): number {
    return Math.ceil(text.length / this.CHARS_PER_TOKEN);
  }
}
